include/rocksdb/table.h

```diff
@@ -51,6 +51,55 @@ enum ChecksumType : char {
   kxxHash64 = 0x3,
 };
 
+// `PinningTier` is used to specify which tier of block-based tables should
+// be affected by a block cache pinning setting (see
+// `MetadataCachePinningOptions` below).
+enum class PinningTier {
+  // For compatibility, this value specifies to fallback to the behavior
+  // indicated by the deprecated options,
+  // `pin_l0_filter_and_index_blocks_in_cache` and
+  // `pin_top_level_index_and_filter`.
+  kFallback,
+
+  // This tier contains no block-based tables.
+  kNone,
+
+  // This tier contains block-based tables that may have originated from a
+  // memtable flush. In particular, it includes tables from L0 that are smaller
+  // than 1.5 times the current `write_buffer_size`. Note these criteria imply
+  // it can include intra-L0 compaction outputs and ingested files, as long as
+  // they are not abnormally large for L0.
+  kMaybeFlushed,
+
+  // This tier contains all block-based tables.
+  kAll,
+};
+
+// `MetadataCachePinningOptions` contains members indicating the desired block
+// cache pinning behavior for the different categories of metadata blocks.
+struct MetadataCachePinningOptions {
+  // The tier of block-based tables whose top-level index into metadata
+  // partitions will be pinned. Currently indexes and filters may be
+  // partitioned.
+  //
+  // Note `cache_index_and_filter_blocks` must be true for this option to have
```

Comment:
Just a question, can `cache_index_and_filter_blocks` also be deprecated or included in this option? Then the option would be like `MetadataCacheOptions`, which has options to enable/disable, pin/unpin the cache.

Reply:
I think that's a great point that could influence how we name these options, even if we don't include it right now.

Reply:
That's a good question. Indeed there's a "TODO(ajkr)" in `cache_index_and_filter_blocks`'s API comment saying we should change its name. `cache_index_and_filter_blocks_with_high_priority` has the same naming problem, i.e., it says it applies to index and filter blocks, while actually it applies to all unpartitioned metadata and top-level indexes into metadata partitions. I am not that sure right now whether we'll want to simply add a new flag that does the same thing under a different name, or do something else, or do nothing. But it might not hurt to change the name to `MetadataCacheOptions` so we're not restricted later... I will make that change for now unless anyone has other suggestions.

Reply:
Done.

Reply:
Having thought a bit more, I think cached vs. not should generally be the same for all metadata blocks. Unbounded is unbounded, regardless of which subset of metadata you for some reason keep bounded despite other metadata left unbounded.

I'm OK with the naming updates, though.